File: 5.binarySearch/69.Sqrt(x)/main.py
# https://leetcode.com/problems/sqrtx/
# TC: O(logn), SC: O(1)
class Solution:
    def mySqrt(self, x: int) -> int:
        l, h = 1, x
        res = 0
        while l <= h:
            m = l + (h - l) // 2
            product = m * m
            # if product == x:
            #     return m
            if product <= x:
                l = m + 1
                res = m
            else:
                h = m - 1
        return res

File: 5.binarySearch/69.Sqrt(x)/main.go
package main

// https://leetcode.com/problems/sqrtx/

// x: 4 -> 2, 16 -> 4
// x: 8 -> 2

// largest <= num
func mySqrt1(x int) int {
	l := 1
	h := x
	res := 0
	for l <= h {
		m := l + (h-l)/2
		square := m * m
		if square == x {
			return m
		}

		if square <= x {
			res = m
			l = m + 1
		} else {
			h = m - 1
		}
	}

	return res
}

// x: 4 -> 2, 16 -> 4
// x: 8 -> 2

// largest <= num
func mySqrt2(x int) int {
	l := 1
	h := x
	for l <= h {
		m := l + (h-l)/2
		square := m * m
		if square == x {
			return m
		}

		if square < x {
			l = m + 1
		} else {
			h = m - 1
		}
	}

    return h
}
